test/controllers/auth/AuthActionSpec.scala

```diff
@@ -0,0 +1,149 @@
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import utils.TestConstants._
+
+class AuthActionSpec extends UnitSpec with OneAppPerSuite with MockitoSugar {
+
+  class BrokenAuthConnector(exception: Throwable) extends AuthConnector {
+    override def authorise[A](predicate: Predicate, retrieval: Retrieval[A])(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[A] =
+      Future.failed(exception)
+  }
+
+  val mockAuthConnector: AuthConnector = mock[AuthConnector]
+
+  class Harness(authAction: AuthAction) extends Controller {
+    def onPageLoad(): Action[AnyContent] = authAction { request => Ok(s"SaUtr: ${request.saUtr.map(_.utr).getOrElse("fail").toString}," +
+      s"AgentRef: ${request.agentRef.map(_.uar).getOrElse("fail").toString}") }
+  }
+
+  val ggSignInUrl = "http://localhost:9025/gg/sign-in?continue=http://localhost:9217/annual-tax-summary&continue=http%3A%2F%2Flocalhost%3A9217%2Fannual-tax-summary&origin=tax-summaries-frontend"
+  implicit val timeout: FiniteDuration = 5 seconds
+
+  "A user with no active session" should {
+    "return 303 and be redirected to GG sign in page" in {
+      when(mockAuthConnector.authorise(any(), any())(any(), any()))
+        .thenReturn(Future.failed(new SessionRecordNotFound))
+      val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+      val controller = new Harness(authAction)
+      val result = controller.onPageLoad()(FakeRequest("", ""))
+      status(result) shouldBe SEE_OTHER
+      redirectLocation(result).get should endWith(ggSignInUrl)
+    }
+  }
+
+  "A user with confidence level below 50" should {
+    "return 303 and be redirected to GG sign in page" in {
+      when(mockAuthConnector.authorise(any(), any())(any(), any()))
+        .thenReturn(Future.failed(new InsufficientConfidenceLevel))
+      val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+      val controller = new Harness(authAction)
+      val result = controller.onPageLoad()(FakeRequest("", ""))
+      status(result) shouldBe SEE_OTHER
+      redirectLocation(result).get should endWith(ggSignInUrl)
+    }
+  }
+
+  "A user with insufficient enrolments" should {
+    "be redirected to the Insufficient Enrolments Page" in {
+      when(mockAuthConnector.authorise(any(), any())(any(), any()))
+        .thenReturn(Future.failed(InsufficientEnrolments()))
+      val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+      val controller = new Harness(authAction)
+      val result = controller.onPageLoad()(FakeRequest("", ""))
+
+      redirectLocation(result) shouldBe Some("/annual-tax-summary/not-authorised")
+    }
+  }
+
+  "A user with a confidence level 50 and an SA enrolment" should {
+    "create an authenticated request" in {
+      val utr = new SaUtrGenerator().nextSaUtr.utr
+      val retrievalResult: Future[
+        Enrolments ~ Option[String] ~ Option[String]] =
+        Future.successful(
+            new ~(
```

Comment:
you can bring in @jaynabonne RetrivalsOps to write this more cleanly

Reply:
Done